File: apps/backend/app/services/experiment/experiment.py
"""
Experiment Service

Business logic for experiment management.
"""

import logging
from typing import Dict, List, Any, Optional
from datetime import datetime

from .mlflow_client import get_mlflow_client

logger = logging.getLogger(__name__)


class ExperimentService:
    """
    Service for managing MLflow experiments
    """

    def __init__(self):
        self._mlflow = get_mlflow_client()

    async def create_experiment(
        self,
        name: str,
        description: Optional[str] = None,
        tags: Optional[Dict[str, str]] = None,
        project_id: Optional[int] = None,
    ) -> Dict[str, Any]:
        """
        Create a new experiment

        Args:
            name: Experiment name
            description: Optional description
            tags: Optional tags
            project_id: Optional project ID for multi-tenancy

        Returns:
            Created experiment metadata
        """
        # Add project tag if provided
        experiment_tags = tags or {}
        if description:
            experiment_tags["description"] = description
        if project_id:
            experiment_tags["project_id"] = str(project_id)

        experiment_id = self._mlflow.create_experiment(
            name=name,
            tags=experiment_tags if experiment_tags else None,
        )

        return await self.get_experiment(experiment_id)

    async def get_experiment(self, experiment_id: str) -> Dict[str, Any]:
        """
        Get experiment by ID

        Args:
            experiment_id: Experiment ID

        Returns:
            Experiment metadata
        """
        exp = self._mlflow.get_experiment(experiment_id)

        # Get run count
        runs = self._mlflow.list_runs(experiment_id, max_results=1)
        run_count = len(self._mlflow.list_runs(experiment_id, max_results=10000))

        # Get best run (by first metric)
        best_run = self._get_best_run(experiment_id)

        return {
            **exp,
            "run_count": run_count,
            "best_run": best_run,
            "description": exp.get("tags", {}).get("description", ""),
        }

    async def list_experiments(
        self,
        project_id: Optional[int] = None,
        search: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """
        List all experiments

        Args:
            project_id: Optional project ID filter
            search: Optional search term for name

        Returns:
            List of experiment metadata with stats
        """
        experiments = self._mlflow.list_experiments()

        # Filter by project if specified
        if project_id is not None:
            experiments = [
                e for e in experiments
                if e.get("tags", {}).get("project_id") == str(project_id)
            ]

        # Filter by search term
        if search:
            search_lower = search.lower()
            experiments = [
                e for e in experiments
                if search_lower in e.get("name", "").lower()
            ]

        # Add run counts and best runs
        result = []
        for exp in experiments:
            runs = self._mlflow.list_runs(exp["id"], max_results=10000)
            run_count = len(runs)

            best_run = None
            if runs:
                # Try to find a run with metrics to determine "best"
                best_run = self._get_best_run(exp["id"])

            result.append({
                **exp,
                "run_count": run_count,
                "best_run": best_run,
                "description": exp.get("tags", {}).get("description", ""),
            })

        return result

    async def update_experiment(
        self,
        experiment_id: str,
        name: Optional[str] = None,
        description: Optional[str] = None,
        tags: Optional[Dict[str, str]] = None,
    ) -> bool:
        """
        Update experiment

        Args:
            experiment_id: Experiment ID
            name: New name
            description: New description (stored as tag)
            tags: New tags (replaces all tags)

        Returns:
            True if updated successfully
        """
        update_tags = tags or {}

        if description:
            update_tags["description"] = description

        return self._mlflow.update_experiment(
            experiment_id=experiment_id,
            name=name,
            tags=update_tags if update_tags else None,
        )

    async def delete_experiment(self, experiment_id: str) -> bool:
        """
        Delete experiment (soft delete)

        Args:
            experiment_id: Experiment ID

        Returns:
            True if deleted successfully
        """
        return self._mlflow.delete_experiment(experiment_id)

    async def restore_experiment(self, experiment_id: str) -> bool:
        """
        Restore deleted experiment

        Args:
            experiment_id: Experiment ID

        Returns:
            True if restored successfully
        """
        return self._mlflow.restore_experiment(experiment_id)

    async def get_experiment_runs(
        self,
        experiment_id: str,
        max_results: int = 100,
        order_by: Optional[List[str]] = None,
    ) -> List[Dict[str, Any]]:
        """
        Get all runs for an experiment

        Args:
            experiment_id: Experiment ID
            max_results: Maximum number of runs
            order_by: Order by clauses

        Returns:
            List of run metadata
        """
        return self._mlflow.list_runs(
            experiment_id=experiment_id,
            max_results=max_results,
            order_by=order_by or ["start_time DESC"],
        )

    async def get_experiment_metrics(
        self,
        experiment_id: str,
        metric_keys: Optional[List[str]] = None,
    ) -> Dict[str, List[Dict[str, Any]]]:
        """
        Get metric history for all metrics in an experiment

        Args:
            experiment_id: Experiment ID
            metric_keys: Optional list of specific metric keys

        Returns:
            Dict mapping metric key to history
        """
        runs = await self.get_experiment_runs(experiment_id, max_results=1000)

        all_metrics: Dict[str, List[Dict[str, Any]]] = {}

        for run in runs:
            run_id = run["run_id"]
            metrics = run.get("metrics", {})

            for key, value in metrics.items():
                if metric_keys and key not in metric_keys:
                    continue

                if key not in all_metrics:
                    all_metrics[key] = []

                all_metrics[key].append({
                    "run_id": run_id,
                    "run_name": run.get("run_name", run_id[:8]),
                    "value": value,
                    "timestamp": run.get("start_time"),
                    "step": 0,  # Static metrics from run summary
                })

        return all_metrics

    async def compare_runs(
        self,
        run_ids: List[str],
    ) -> Dict[str, Any]:
        """
        Compare multiple runs

        Args:
            run_ids: List of run IDs to compare

        Returns:
            Comparison data with params and metrics
        """
        runs = []
        all_params = set()
        all_metrics = set()

        for run_id in run_ids:
            run = self._mlflow.get_run(run_id)
            runs.append(run)
            all_params.update(run.get("params", {}).keys())
            all_metrics.update(run.get("metrics", {}).keys())

        return {
            "runs": runs,
            "param_keys": sorted(all_params),
            "metric_keys": sorted(all_metrics),
        }

    async def delete_runs_batch(
        self,
        experiment_id: str,
        run_ids: List[str],
    ) -> Dict[str, bool]:
        """
        Delete multiple runs

        Args:
            experiment_id: Experiment ID
            run_ids: List of run IDs to delete

        Returns:
            Dict mapping run_id to success status
        """
        results = {}
        for run_id in run_ids:
            results[run_id] = self._mlflow.delete_run(run_id)
        return results

    def _get_best_run(
        self,
        experiment_id: str,
        metric_key: str = None,
        mode: str = "max",
    ) -> Optional[Dict[str, Any]]:
        """
        Get the best run for an experiment

        Args:
            experiment_id: Experiment ID
            metric_key: Metric key to optimize (uses first available if None)
            mode: "min" or "max"

        Returns:
            Best run metadata or None
        """
        runs = self._mlflow.list_runs(experiment_id, max_results=1000)

        if not runs:
            return None

        # Filter runs with metrics
        runs_with_metrics = [r for r in runs if r.get("metrics")]
        if not runs_with_metrics:
            return None

        # Determine metric to use
        if metric_key is None:
            # Use first available metric
            first_run = runs_with_metrics[0]
            metric_key = list(first_run.get("metrics", {}).keys())[0]

        # Find best run
        best_run = None
        best_value = None

        for run in runs_with_metrics:
            value = run.get("metrics", {}).get(metric_key)
            if value is None:
                continue

            if best_value is None:
                best_value = value
                best_run = run
            elif mode == "max" and value > best_value:
                best_value = value
                best_run = run
            elif mode == "min" and value < best_value:
                best_value = value
                best_run = run

        if best_run:
            return {
                **best_run,
                "optimized_metric": metric_key,
                "optimized_value": best_value,
            }

        return None


# Singleton instance
_experiment_service: Optional[ExperimentService] = None


def get_experiment_service() -> ExperimentService:
    """Get or create experiment service singleton"""
    global _experiment_service
    if _experiment_service is None:
        _experiment_service = ExperimentService()
    return _experiment_service

File: apps/frontend/src/stores/experiment.ts
/**
 * Experiment Store
 *
 * Zustand store for managing MLflow experiment state and API calls.
 */

import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import api from '@/services/api';

// Types
export interface Experiment {
  id: string;
  name: string;
  description?: string;
  artifact_location?: string;
  lifecycle_stage: string;
  tags?: Record<string, string>;
  created_at: string;
  updated_at?: string;
  run_count?: number;
  best_run?: Run;
}

export interface Run {
  run_id: string;
  run_uuid: string;
  run_name?: string;
  experiment_id: string;
  status: 'running' | 'completed' | 'failed' | 'killed' | 'scheduled';
  lifecycle_stage: string;
  start_time?: number;
  end_time?: number;
  artifact_uri?: string;
  params?: Record<string, string>;
  metrics?: Record<string, number>;
  tags?: Record<string, string>;
  artifacts?: Artifact[];
}

export interface Artifact {
  path: string;
  is_dir: boolean;
  size?: number;
}

export interface MetricHistory {
  key: string;
  value: number;
  timestamp: number;
  step: number;
}

export interface MetricComparison {
  runs: Run[];
  param_keys: string[];
  metric_keys: string[];
  metric_stats?: Record<string, {
    values: Array<{
      run_id: string;
      run_name: string;
      value: number;
    }>;
    min: number;
    max: number;
    mean: number;
    median: number;
    stdev: number;
  }>;
}

interface ExperimentState {
  // State
  experiments: Experiment[];
  currentExperiment: Experiment | null;
  runs: Run[];
  currentRun: Run | null;
  loading: boolean;
  error: string | null;

  // Experiment actions
  fetchExperiments: (projectId?: string, search?: string) => Promise<void>;
  fetchExperiment: (experimentId: string) => Promise<void>;
  createExperiment: (data: {
    name: string;
    description?: string;
    tags?: Record<string, string>;
    projectId?: number;
  }) => Promise<Experiment>;
  updateExperiment: (
    experimentId: string,
    data: { name?: string; description?: string; tags?: Record<string, string> }
  ) => Promise<void>;
  deleteExperiment: (experimentId: string) => Promise<void>;
  restoreExperiment: (experimentId: string) => Promise<void>;

  // Run actions
  fetchRuns: (experimentId: string, maxResults?: number, status?: string) => Promise<void>;
  fetchRun: (runId: string) => Promise<void>;
  createRun: (data: {
    experimentId: string;
    runName?: string;
    tags?: Record<string, string>;
  }) => Promise<Run>;
  deleteRun: (runId: string) => Promise<void>;

  // Logging actions
  logParams: (runId: string, params: Record<string, any>) => Promise<void>;
  logMetrics: (runId: string, metrics: Record<string, number>, step?: number) => Promise<void>;
  logBatch: (
    runId: string,
    data: {
      params?: Record<string, any>;
      metrics?: Record<string, number>;
      tags?: Record<string, string>;
    }
  ) => Promise<void>;
  setRunStatus: (runId: string, status: 'FINISHED' | 'FAILED' | 'KILLED') => Promise<void>;

  // Metric actions
  getMetricHistory: (runId: string, metricKey: string) => Promise<MetricHistory[]>;
  compareMetrics: (runIds: string[], metricKeys?: string[]) => Promise<MetricComparison>;
  getMetricSummary: (experimentId: string, metricKey?: string) => Promise<any>;
  getMetricCorrelation: (experimentId: string, metricKey: string) => Promise<any>;

  // Comparison actions
  compareRuns: (runIds: string[]) => Promise<MetricComparison>;

  // Artifact actions
  listArtifacts: (runId: string, path?: string) => Promise<Artifact[]>;

  // State management
  setCurrentExperiment: (experiment: Experiment | null) => void;
  setCurrentRun: (run: Run | null) => void;
  clearError: () => void;
}

export const useExperimentStore = create<ExperimentState>()(
  persist(
    (set, get) => ({
      // Initial state
      experiments: [],
      currentExperiment: null,
      runs: [],
      currentRun: null,
      loading: false,
      error: null,

      // Experiment actions
      fetchExperiments: async (projectId?: string, search?: string) => {
        set({ loading: true, error: null });
        try {
          const params: Record<string, string> = {};
          if (projectId) params.project_id = String(projectId);
          if (search) params.search = search;

          const response = await api.get('/experiments', { params });
          set({ experiments: response.data, loading: false });
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to fetch experiments',
            loading: false,
          });
          throw error;
        }
      },

      fetchExperiment: async (experimentId: string) => {
        set({ loading: true, error: null });
        try {
          const response = await api.get(`/experiments/${experimentId}`);
          set({ currentExperiment: response.data, loading: false });
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to fetch experiment',
            loading: false,
          });
          throw error;
        }
      },

      createExperiment: async (data) => {
        set({ loading: true, error: null });
        try {
          const response = await api.post('/experiments', data);
          set((state) => ({
            experiments: [...state.experiments, response.data],
            loading: false,
          }));
          return response.data;
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to create experiment',
            loading: false,
          });
          throw error;
        }
      },

      updateExperiment: async (experimentId, data) => {
        set({ loading: true, error: null });
        try {
          await api.put(`/experiments/${experimentId}`, data);
          set((state) => ({
            experiments: state.experiments.map((exp) =>
              exp.id === experimentId ? { ...exp, ...data } : exp
            ),
            currentExperiment: state.currentExperiment?.id === experimentId
              ? { ...state.currentExperiment, ...data }
              : state.currentExperiment,
            loading: false,
          }));
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to update experiment',
            loading: false,
          });
          throw error;
        }
      },

      deleteExperiment: async (experimentId: string) => {
        set({ loading: true, error: null });
        try {
          await api.delete(`/experiments/${experimentId}`);
          set((state) => ({
            experiments: state.experiments.filter((exp) => exp.id !== experimentId),
            loading: false,
          }));
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to delete experiment',
            loading: false,
          });
          throw error;
        }
      },

      restoreExperiment: async (experimentId: string) => {
        set({ loading: true, error: null });
        try {
          await api.post(`/experiments/${experimentId}/restore`);
          await get().fetchExperiments();
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to restore experiment',
            loading: false,
          });
          throw error;
        }
      },

      // Run actions
      fetchRuns: async (experimentId: string, maxResults = 100, status?: string) => {
        set({ loading: true, error: null });
        try {
          const params: Record<string, string | number> = { max_results: maxResults };
          if (status) params.status = status;

          const response = await api.get(`/experiments/${experimentId}/runs`, { params });
          set({ runs: response.data, loading: false });
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to fetch runs',
            loading: false,
          });
          throw error;
        }
      },

      fetchRun: async (runId: string) => {
        set({ loading: true, error: null });
        try {
          const response = await api.get(`/experiments/runs/${runId}`);
          set({ currentRun: response.data, loading: false });
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to fetch run',
            loading: false,
          });
          throw error;
        }
      },

      createRun: async (data) => {
        set({ loading: true, error: null });
        try {
          const response = await api.post('/experiments/runs', data);
          set((state) => ({
            runs: [response.data, ...state.runs],
            loading: false,
          }));
          return response.data;
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to create run',
            loading: false,
          });
          throw error;
        }
      },

      deleteRun: async (runId: string) => {
        set({ loading: true, error: null });
        try {
          await api.delete(`/experiments/runs/${runId}`);
          set((state) => ({
            runs: state.runs.filter((run) => run.run_id !== runId),
            loading: false,
          }));
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to delete run',
            loading: false,
          });
          throw error;
        }
      },

      // Logging actions
      logParams: async (runId: string, params: Record<string, any>) => {
        set({ loading: true, error: null });
        try {
          await api.post(`/experiments/runs/${runId}/params`, { params });
          set({ loading: false });
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to log params',
            loading: false,
          });
          throw error;
        }
      },

      logMetrics: async (runId: string, metrics: Record<string, number>, step = 0) => {
        set({ loading: true, error: null });
        try {
          await api.post(`/experiments/runs/${runId}/metrics`, { metrics, step });
          set({ loading: false });
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to log metrics',
            loading: false,
          });
          throw error;
        }
      },

      logBatch: async (runId, data) => {
        set({ loading: true, error: null });
        try {
          await api.post(`/experiments/runs/${runId}/batch`, data);
          set({ loading: false });
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to log batch',
            loading: false,
          });
          throw error;
        }
      },

      setRunStatus: async (runId, status) => {
        set({ loading: true, error: null });
        try {
          await api.post(`/experiments/runs/${runId}/status`, null, {
            params: { run_status: status },
          });
          set((state) => ({
            runs: state.runs.map((run) =>
              run.run_id === runId ? { ...run, status: status.toLowerCase() as any } : run
            ),
            loading: false,
          }));
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to set run status',
            loading: false,
          });
          throw error;
        }
      },

      // Metric actions
      getMetricHistory: async (runId: string, metricKey: string) => {
        try {
          const response = await api.get(`/experiments/runs/${runId}/metrics/history`, {
            params: { metric_key: metricKey },
          });
          return response.data.history;
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to get metric history',
          });
          throw error;
        }
      },

      compareMetrics: async (runIds: string[], metricKeys?: string[]) => {
        try {
          const response = await api.post('/experiments/metrics/compare', {
            run_ids: runIds,
            metric_keys: metricKeys,
          });
          return response.data;
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to compare metrics',
          });
          throw error;
        }
      },

      getMetricSummary: async (experimentId: string, metricKey?: string) => {
        try {
          const params: Record<string, string> = {};
          if (metricKey) params.metric_key = metricKey;

          const response = await api.get(`/experiments/${experimentId}/metrics/summary`, {
            params,
          });
          return response.data;
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to get metric summary',
          });
          throw error;
        }
      },

      getMetricCorrelation: async (experimentId: string, metricKey: string) => {
        try {
          const response = await api.get(`/experiments/${experimentId}/metrics/correlation`, {
            params: { metric_key: metricKey },
          });
          return response.data;
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to get metric correlation',
          });
          throw error;
        }
      },

      // Comparison actions
      compareRuns: async (runIds: string[]) => {
        try {
          const response = await api.post('/experiments/runs/compare', runIds);
          return response.data;
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to compare runs',
          });
          throw error;
        }
      },

      // Artifact actions
      listArtifacts: async (runId: string, path?: string) => {
        try {
          const params: Record<string, string> = {};
          if (path) params.path = path;

          const response = await api.get(`/experiments/runs/${runId}/artifacts`, { params });
          return response.data.artifacts;
        } catch (error: any) {
          set({
            error: error.response?.data?.detail || 'Failed to list artifacts',
          });
          throw error;
        }
      },

      // State management
      setCurrentExperiment: (experiment) => set({ currentExperiment: experiment }),
      setCurrentRun: (run) => set({ currentRun: run }),
      clearError: () => set({ error: null }),
    }),
    {
      name: 'experiment-store',
      partialize: (state) => ({
        experiments: state.experiments,
      }),
    }
  )
);

// Selectors
export const useExperiments = () => useExperimentStore((state) => state.experiments);
export const useCurrentExperiment = () => useExperimentStore((state) => state.currentExperiment);
export const useRuns = () => useExperimentStore((state) => state.runs);
export const useCurrentRun = () => useExperimentStore((state) => state.currentRun);
export const useExperimentLoading = () => useExperimentStore((state) => state.loading);
export const useExperimentError = () => useExperimentStore((state) => state.error);
